// threadedclass/dist/fakeProcess.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const internalApi_1 = require("./internalApi");
const workerPlatformBase_1 = require("./workerPlatformBase");
class FakeWorker extends internalApi_1.Worker {
    constructor(cb) {
        super();
        this.disabledMultithreading = true;
        this.mockProcessSend = cb;
    }
    killInstance() {
        // throw new Error('Trying to kill a non threaded process!')
    }
    sendMessageToParent(handle, msg, cb) {
        if (msg.cmd === internalApi_1.MessageType.LOG) {
            const message = Object.assign(Object.assign({}, msg), {
                cmdId: 0,
                instanceId: ''
            });
            // Send message to Parent:
            this.mockProcessSend(message);
        }
        else {
            const message = Object.assign(Object.assign({}, msg), {
                cmdId: handle.cmdId++,
                instanceId: handle.id
            });
            if (cb)
                handle.queue[message.cmdId + ''] = cb;
            // Send message to Parent:
            this.mockProcessSend(message);
        }
    }
}
exports.FakeWorker = FakeWorker;
class FakeProcess extends workerPlatformBase_1.WorkerPlatformBase {
    constructor() {
        super();
        this.worker = new FakeWorker((m) => {
            this.emit('message', m);
        });
    }
    kill() {
        // @todo: needs some implementation.
        this.emit('close');
    }
    send(m) {
        this.worker.onMessageFromParent(m);
    }
}
exports.FakeProcess = FakeProcess;
